### PR TITLE
removing any remaining need for the alignment app

### PR DESCRIPTION
**Pull request by railsagainstignorance**
new podcast url: /podcastBasedOnFirstFt/:maxResults/:voice?token=TOKEN
or /podcastBasedOnFirstFt/:maxResults/:voice?token=TOKEN&skipFirstFtUuids=true

includes any additionally specified UUIDs

with assorted partial/test endpoints via index page

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,6 +13,8 @@ More details
 
 # new end points
 
+* /podcastBasedOnFirstFt/:maxResults/:voice?token=TOKEN
+* /podcastBasedOnFirstFt/:maxResults/:voice?token=TOKEN&skipFirstFtUuids=true
```
Comment (seanmtracey) [inline README.md]: What affect does `skipFirstFtUuids=true` have?

File: bin/lib/autovoice.js [hunk 2]
```diff
@@ -98,6 +99,27 @@ function generatePodcast(rssUrl, voiceId=tts.defaultVoiceId){
...
+	return fetchContent.getLastFewFirstFtMentionedUuids(maxResults, includeFirstFtUuids)
+	.then( firstFtBasedUuids => firstFtBasedUuids.concat(individualUUIDs.list()) )
+	.then( uuids => fetchContent.articlesAsItems( uuids ) )
+	.then( items => {
+		debug(`generateFirstFtBasedPodcast: items.length=${items.length}`);
+		const promises = items.map( (item, i) => {
+			item.processingIndex = i;
+			return processItemToMp3(item, voiceId);
+		} );
+		return Promise.all(promises);
```
Comment (seanmtracey) [inline bin/lib/autovoice.js]: None of these promises are being caught if an error occurs. All of them will be rejected if one of the set fails.

File: bin/lib/autovoice.js [hunk 3]
```diff
@@ -98,6 +99,27 @@ function generatePodcast(rssUrl, voiceId=tts.defaultVoiceId){
...
+	})
+	.then( items => {
+		const feed = constructRSS(requestedUrl, items);
+		return feed;
```
Comment (seanmtracey) [inline bin/lib/autovoice.js]: Why not just `return constructRSS(requestedUrl, items)`?

File: bin/lib/fetchContent.js [hunk 4]
```diff
@@ -1,18 +1,128 @@
...
+		     offset : 0,
+		    aspects : [ "title"], // [ "title", "location", "summary", "lifecycle", "metadata"],
+		constraints : []
+	};
+
+	const combined = Object.assign({}, defaults, params);
+
+	let queryString = combined.queryString;
+	if (queryString == '' && combined.constraints.length > 0 ) {
+		queryString = combined.constraints.join(' and ');
```
Comment (seanmtracey) [inline bin/lib/fetchContent.js]: The SAPI query appears to be constructed from query parameters. Are you doing any kind of sanitising of the input?

File: bin/lib/fetchContent.js [hunk 5]
```diff
@@ -1,18 +1,128 @@
...
+}
+
+function search(params) {
+	const sapiUrl = `${SAPI_PATH}?apiKey=${CAPI_KEY}`;
+	const sapiQuery = constructSAPIQuery( params );
+	debug(`search: sapiQuery=${JSON.stringify(sapiQuery)}`);
+
+	return fetch(sapiUrl, {
+		 method: 'POST',
+       body: JSON.stringify(sapiQuery),
```
Comment (seanmtracey) [inline bin/lib/fetchContent.js]: Unusual indenting.

File: bin/lib/fetchContent.js [hunk 6]
```diff
@@ -1,18 +1,128 @@
...
+		headers: {
+			'Content-Type' : 'application/json',
+		},
+	})
+	.then( res  => res.text() )
```
Comment (seanmtracey) [inline bin/lib/fetchContent.js]: You're returning this as text, and then your parsing it as JSON on line 72. You can have the response handled and parsed as JSON automatically.

```
// Replacement for lines 64 - 76
.then( res => res.json() )
.then(data => {
	debug(`search: res.json=${JSON.stringify(data)}`);
	return {
 		params : params,
 		sapiObj : data
	};
})
```

File: bin/lib/fetchContent.js [hunk 7]
```diff
@@ -1,18 +1,128 @@
...
+	return search({queryString: uuid});
+}
+
+function searchLastFewFirstFt(maxResults) {
+	return search({queryString: `brand:FirstFT`, maxResults: maxResults});
+}
+
+function extractFirstFtIds( sapiObj ){
+	let uuids = [];
+	if (! sapiObj.results ) {
```
Comment (seanmtracey) [inline bin/lib/fetchContent.js]: Should these `throw` an error to be handled, or is it ok to return an empty array if something is amiss?

File: bin/lib/fetchContent.js [hunk 8]
```diff
@@ -1,18 +1,128 @@
...
+	} else {
+		uuids = sapiObj.results[0].results.map( r => { return r.id; } );
+	}
+	return uuids
+}
+
+// <ft-content type=\"http://www.ft.com/ontology/content/Article\" url=\"http://api.ft.com/content/dd033082-49e9-11e7-a3f4-c742b9791d43\" title=\"www.ft.com\">paid up to $1bn</ft-content>
+const reFtContent = new RegExp(/<ft-content\s+type=\"http:\/\/www.ft.com\/ontology\/content\/Article\"\s+url=\"http:\/\/api.ft.com\/content\/([a-f0-9-]+)\"/, 'g');
+
+function getLastFewFirstFtMentionedUuids(maxResults, includeFirstFtUuids=false) {
```
Comment (seanmtracey) [inline bin/lib/fetchContent.js]: This function doesn't `catch` errors. If only one of the promises in `Promise.all` fails, the whole function will fail. Are errors being handled somewhere further up the chain?

File: index.js [hunk 9]
```diff
@@ -52,6 +52,26 @@ app.get('/podcast', (req, res) => {
...
 
+app.get('/podcastBasedOnFirstFt/:maxResults/:voice', (req, res) => {
+  const   maxResults = req.params.maxResults;
+  const       voice  = req.params.voice;
+  const requestedUrl = process.env.SERVER_ROOT + '/' + req.originalUrl;
+  const       token  = req.query.token;
+  const skipFirstFtUuids = req.query.skipFirstFtUuids;
+  const includeFirstFtUuids = !(skipFirstFtUuids == 'true');
+
+  if       ( ! token                 ) { res.status(400).send('This call requires a token parameter.'      ).end();
```
Comment (seanmtracey) [inline index.js]: What's going on with all of these spaces?

File: index.js [hunk 10]
```diff
@@ -52,6 +52,26 @@ app.get('/podcast', (req, res) => {
...
+  } else if( token !== PODCAST_TOKEN ) { res.status(401).send('This call requires a valid token parameter.').end();
```
Comment (seanmtracey) [inline index.js]: You don't need `.end()` if you're using `.send()`.

File: static/index.html [hunk 11]
```diff
@@ -7,7 +7,9 @@
   <body>
     <h1>Assorted endpoints for autovoice</h1>
     <ul>
-      <li>/podcast?rss=URL_OF_ARTICLE_RSS_FEED</li>
+      <li>/podcast?voice=VOICE&token=TOKEN&rss=URL_OF_ARTICLE_RSS_FEED</li>
```
Comment (seanmtracey) [inline static/index.html]: `&` should be `&amp;`

File: README.md (hunk 1 above)
Comment (railsagainstignorance) [inline README.md]: it ensures that only the FT articles referred to by firstFT are voiced, and not the actual firstFT articles themselves

File: bin/lib/autovoice.js (hunk 2 above)
Comment (railsagainstignorance) [inline bin/lib/autovoice.js]: looking for suggestions for sensible error handling

File: bin/lib/autovoice.js (hunk 3 above)
Comment (railsagainstignorance) [inline bin/lib/autovoice.js]: no good reason.

File: bin/lib/fetchContent.js (hunk 7 above)
Comment (railsagainstignorance) [inline bin/lib/fetchContent.js]: empty array should be fine.

File: README.md (hunk 1 above)
Comment (railsagainstignorance) [inline README.md]: amended README

File: bin/lib/autovoice.js (hunk 3 above)
Comment (railsagainstignorance) [inline bin/lib/autovoice.js]: replaced with 
	.then( items => constructRSS(requestedUrl, items) )

File: bin/lib/fetchContent.js (hunk 4 above)
Comment (railsagainstignorance) [inline bin/lib/fetchContent.js]: the fn constructSAPIQuery which params are being looked for and why, but does not sanitise them.

File: bin/lib/fetchContent.js (hunk 5 above)
Comment (railsagainstignorance) [inline bin/lib/fetchContent.js]: LGTM. Might be a tabs problem.

File: index.js (hunk 10 above)
Comment (railsagainstignorance) [inline index.js]: done

File: static/index.html (hunk 11 above)
Comment (railsagainstignorance) [inline static/index.html]: done

**Comment by railsagainstignorance** [comment]
fixed #7

**Pull request merged**